## Subash Chandra Biswal (U77884251) ##
# Assignment 1 - Cardiotocography


## Introduction and Overview


Author: J. P. Marques de Sá, J. Bernardes, D. Ayers de Campos.  
Source: UCI  
Please cite: Ayres de Campos et al. (2000) SisPorto 2.0 A Program for Automated Analysis of Cardiotocograms. J Matern Fetal Med 5:311-318, UCI    

2126 fetal cardiotocograms (CTGs) were automatically processed and the respective diagnostic features measured. The CTGs were also classified by three expert obstetricians and a consensus classification label assigned to each of them. Classification was both with respect to a morphologic pattern (A, B, C. ...) and to a fetal state (N, S, P). Therefore the dataset can be used either for 10-class or 3-class experiments.  

Attribute Information:  
LB - FHR baseline (beats per minute)  
AC - # of accelerations per second  
FM - # of fetal movements per second  
UC - # of uterine contractions per second  
DL - # of light decelerations per second  
DS - # of severe decelerations per second  
DP - # of prolongued decelerations per second  
ASTV - percentage of time with abnormal short term variability  
MSTV - mean value of short term variability  
ALTV - percentage of time with abnormal long term variability  
MLTV - mean value of long term variability  
Width - width of FHR histogram  
Min - minimum of FHR histogram  
Max - Maximum of FHR histogram  
Nmax - # of histogram peaks  
Nzeros - # of histogram zeros  
Mode - histogram mode  
Mean - histogram mean  
Median - histogram median  
Variance - histogram variance  
Tendency - histogram tendency  
CLASS - FHR pattern class code (1 to 10)  
NSP - fetal state class code (N=normal(1); S=suspect(2); P=pathologic(3))  

## Install and import necessary packages

In [1]:
# import packages
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

from sklearn import preprocessing
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder

from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier 
from sklearn.svm import SVC
from xgboost import XGBClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, AdaBoostClassifier

# set random seed to ensure that results are repeatable
np.random.seed(1)

## Load data 

In [2]:
df = pd.read_csv('./cardiotocography_csv.csv')
df.head(5)

V1   V2    V3   V4   V5  V6  V7  V8  V9  V10  ...  V28  V29  V30  V31  V32  \
0  23  240   357  120  120   0   0   0  73  0.5  ...    0    0    0    0    0   
1  45    5   632  132  132   4   0   4  17  2.1  ...    0    0    0    1    0   
2  45  177   779  133  133   2   0   5  16  2.1  ...    0    0    0    1    0   
3  45  411  1192  134  134   2   0   6  16  2.4  ...    0    0    0    1    0   
4  45  533  1147  132  132   4   0   5  16  2.4  ...    0    0    0    0    0   

   V33  V34  V35  Class  NSP  
0    0    1    0      9    2  
1    0    0    0      6    1  
2    0    0    0      6    1  
3    0    0    0      6    1  
4    0    0    0      2    1  

[5 rows x 37 columns]

## Explore the dataset

In [3]:
# Explore the dataset
# read the first row of the dataset 
print(df.head())
print(df.columns)
print(df.describe())
print(df.info())

   V1   V2    V3   V4   V5  V6  V7  V8  V9  V10  ...  V28  V29  V30  V31  V32  \
0  23  240   357  120  120   0   0   0  73  0.5  ...    0    0    0    0    0   
1  45    5   632  132  132   4   0   4  17  2.1  ...    0    0    0    1    0   
2  45  177   779  133  133   2   0   5  16  2.1  ...    0    0    0    1    0   
3  45  411  1192  134  134   2   0   6  16  2.4  ...    0    0    0    1    0   
4  45  533  1147  132  132   4   0   5  16  2.4  ...    0    0    0    0    0   

   V33  V34  V35  Class  NSP  
0    0    1    0      9    2  
1    0    0    0      6    1  
2    0    0    0      6    1  
3    0    0    0      6    1  
4    0    0    0      2    1  

[5 rows x 37 columns]
Index(['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11',
       'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21',
       'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'V29', 'V30', 'V31',
       'V32', 'V33', 'V34', 'V35', 'Class', 'NSP'],
      dtype='object')
 

## Clean/transform data (where necessary)

In [4]:
# based on findings from data exploration, we need to clean up colum names, as there are some leading whitespace characters
df.columns = [s.strip() for s in df.columns] 
df.columns

Index(['V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10', 'V11',
       'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20', 'V21',
       'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'V29', 'V30', 'V31',
       'V32', 'V33', 'V34', 'V35', 'Class', 'NSP'],
      dtype='object')

Drop the columns we are not using as predictors (see previous notebooks -- we are given a subset of input variables to consider). The Class variable is a target variable and has 10 classes. But, we are using the NSP variable as our target variable. So we can drop the Class variable.

In [5]:
df = df.drop(columns=['Class'])

The V25 variable has 3 categories such as 0, 1 and 2. We can encode this variable,

In [6]:
# translation V25 categories into dummy vars
df = df.join(pd.get_dummies(df['V25'], prefix='V25', drop_first=True))
df.drop('V25', axis=1, inplace = True)

df.head(3)

V1   V2   V3   V4   V5  V6  V7  V8  V9  V10  ...  V29  V30  V31  V32  V33  \
0  23  240  357  120  120   0   0   0  73  0.5  ...    0    0    0    0    0   
1  45    5  632  132  132   4   0   4  17  2.1  ...    0    0    1    0    0   
2  45  177  779  133  133   2   0   5  16  2.1  ...    0    0    1    0    0   

   V34  V35  NSP  V25_0  V25_1  
0    1    0    2      0      1  
1    0    0    1      1      0  
2    0    0    1      1      0  

[3 rows x 37 columns]

In [7]:
df.isna().sum()

V1       0
V2       0
V3       0
V4       0
V5       0
V6       0
V7       0
V8       0
V9       0
V10      0
V11      0
V12      0
V13      0
V14      0
V15      0
V16      0
V17      0
V18      0
V19      0
V20      0
V21      0
V22      0
V23      0
V24      0
V26      0
V27      0
V28      0
V29      0
V30      0
V31      0
V32      0
V33      0
V34      0
V35      0
NSP      0
V25_0    0
V25_1    0
dtype: int64

The NSP target variable has 3 classes such as Normal (1), Suspect(2), and Pathologic (3). Out of these 3 classes suspect and pathologic are the stage of concerns and need immediate intervention to control the symptoms. So we can combine suspect and pathologic as the suspect class to target on these conditions. We are assigning 1 to suspect/Pathologic and 0 to Normal stage.

In [8]:
df['NSP'] = np.where(df['NSP']==1, 0, 1)

In [9]:
df.head(5)

V1   V2    V3   V4   V5  V6  V7  V8  V9  V10  ...  V29  V30  V31  V32  V33  \
0  23  240   357  120  120   0   0   0  73  0.5  ...    0    0    0    0    0   
1  45    5   632  132  132   4   0   4  17  2.1  ...    0    0    1    0    0   
2  45  177   779  133  133   2   0   5  16  2.1  ...    0    0    1    0    0   
3  45  411  1192  134  134   2   0   6  16  2.4  ...    0    0    1    0    0   
4  45  533  1147  132  132   4   0   5  16  2.4  ...    0    0    0    0    0   

   V34  V35  NSP  V25_0  V25_1  
0    1    0    1      0      1  
1    0    0    0      1      0  
2    0    0    0      1      0  
3    0    0    0      0      1  
4    0    0    0      0      1  

[5 rows x 37 columns]

## Split data intro training and validation sets

In [10]:
# construct datasets for analysis
target = 'NSP'
predictors = list(df.columns)
predictors.remove(target)
X = df[predictors]
y = df[target]

In [11]:
# create the training set and the test set 
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.25, random_state=1)

Since this is pharmacutical data and we are targeting suspects from various medical test data, we need to minimize the false negatives as this will cost somebody's life. This cost is significantly high as compared to false positive and in case of false positive the patient/insurance company needs to bear only the further investigation costs. So our measure of score is recall.

In [12]:
score_measure = "recall"
kfolds = 5

dtree = DecisionTreeClassifier()
svmm = SVC()
logreg = LogisticRegression()
adatree = AdaBoostClassifier()
rforest = RandomForestClassifier()
xgboost = XGBClassifier()
gboost = GradientBoostingClassifier()

##  Random search of parameter grids of all models ##

In [18]:
#Grid for Logistic Regression

param_grid_logr = [{
     'penalty': ['l1', 'l2', 'elasticnet', 'none'],
     'solver': ['saga'],
     'max_iter': np.arange(100,900),},
      {
     'penalty': ['l1', 'l2'],
     'solver': ['liblinear'],
     'max_iter': np.arange(100,900),},
    {
     'penalty': ['l2', 'none'],
     'solver': ['lbfgs'],
     'max_iter': np.arange(100,900),}
]    

rand_search_logr = RandomizedSearchCV(estimator = logreg, param_distributions=param_grid_logr, cv=kfolds, n_iter=100,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# Logistic Regression model fit for grid search
_ = rand_search_logr.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search_logr.best_score_}")
print(f"... with parameters: {rand_search_logr.best_params_}")

bestRecallLogr = rand_search_logr.best_estimator_

Fitting 5 folds for each of 100 candidates, totalling 500 fits
The best recall score is 0.9539958592132505
... with parameters: {'solver': 'liblinear', 'penalty': 'l1', 'max_iter': 200}


C:\Users\scbis\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
85 fits failed out of a total of 500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
85 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\scbis\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\scbis\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py", line 1471, in fit
    raise ValueError(
ValueError: l1_ratio must be between 0 and 1; got (l1_ratio=None)

  warnings.warn(some_fits_failed_message, FitFailedWarning)
C:\Users\scbis\anaconda3\lib\site-packages\sklearn\mode

In [19]:
# Grid for decision tree
param_grid_tree = {
    'min_samples_split': np.arange(1,100),  
    'min_samples_leaf': np.arange(1,100),
    'min_impurity_decrease': np.arange(0.0001, 0.01, 0.0005),
    'max_leaf_nodes': np.arange(5, 50), 
    'max_depth': np.arange(1,50), 
    'criterion': ['entropy', 'gini'],
}

rand_search_tree = RandomizedSearchCV(estimator = dtree, param_distributions=param_grid_tree, cv=kfolds, n_iter=100,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# Decision tree model fit for grid search
_ = rand_search_tree.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search_tree.best_score_}")
print(f"... with parameters: {rand_search_tree.best_params_}")

bestRecallTree = rand_search_tree.best_estimator_

Fitting 5 folds for each of 100 candidates, totalling 500 fits
The best recall score is 0.9453416149068324
... with parameters: {'min_samples_split': 31, 'min_samples_leaf': 4, 'min_impurity_decrease': 0.0021, 'max_leaf_nodes': 43, 'max_depth': 15, 'criterion': 'entropy'}


C:\Users\scbis\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
5 fits failed out of a total of 500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
5 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\scbis\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\scbis\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 937, in fit
    super().fit(
  File "C:\Users\scbis\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 250, in fit
    raise ValueError(
ValueError: min_samples_split must be an integer greater than 1 or a float in (0.0

In [17]:
# Grid for SVM
param_grid_svm = [{
    'degree': [2,3],
    'C': [1,5,10],
    'kernel': ['poly'],   
},
{
    'C': [1,5,10],
    'gamma': [1, 0.1],
    'kernel': ['rbf'],   
},
{
    'C': [1,5,10],
    'kernel': ['linear'],  
}]

rand_search_svm = RandomizedSearchCV(estimator = svmm, param_distributions=param_grid_svm, cv=kfolds, n_iter=50,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# SVM model fit for grid search
_ = rand_search_svm.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search_svm.best_score_}")
print(f"... with parameters: {rand_search_svm.best_params_}")

bestRecallSvm = rand_search_svm.best_estimator_

C:\Users\scbis\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 15 is smaller than n_iter=50. Running 15 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 5 folds for each of 15 candidates, totalling 75 fits
The best recall score is 0.9712629399585921
... with parameters: {'kernel': 'linear', 'C': 1}


In [20]:
#Grid for ADABoost Classifier

param_grid_ada = {  
     'n_estimators': [10,50,250,1000,2000],
     'learning_rate': [0.01,0.1,0.2,1.0],}   

rand_search_ada = RandomizedSearchCV(estimator = adatree, param_distributions=param_grid_ada, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# ADABoost Classifier model fit for grid search
_ = rand_search_ada.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search_ada.best_score_}")
print(f"... with parameters: {rand_search_ada.best_params_}")

bestRecallAda = rand_search_ada.best_estimator_

C:\Users\scbis\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 20 is smaller than n_iter=500. Running 20 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 5 folds for each of 20 candidates, totalling 100 fits
The best recall score is 0.959792960662526
... with parameters: {'n_estimators': 1000, 'learning_rate': 0.1}


In [21]:
#Grid for Randomforest Classifier

param_grid_rf = {  
     'n_estimators': [10,50,250,1000,2000],
     'max_features': ['auto', 'sqrt', 'log2'],
     'max_depth' : [4,6,8,10],
     'criterion' :['gini', 'entropy', 'log_loss'],}   

rand_search_rf = RandomizedSearchCV(estimator = rforest, param_distributions=param_grid_rf, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# ADABoost Classifier model fit for grid search
_ = rand_search_rf.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search_rf.best_score_}")
print(f"... with parameters: {rand_search_rf.best_params_}")

bestRecallRf = rand_search_rf.best_estimator_

C:\Users\scbis\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 180 is smaller than n_iter=500. Running 180 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 5 folds for each of 180 candidates, totalling 900 fits


C:\Users\scbis\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
300 fits failed out of a total of 900.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
300 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\scbis\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\scbis\anaconda3\lib\site-packages\sklearn\ensemble\_forest.py", line 450, in fit
    trees = Parallel(
  File "C:\Users\scbis\anaconda3\lib\site-packages\joblib\parallel.py", line 1043, in __call__
    if self.dispatch_one_batch(iterator):
  File "C:\Users\scbis\anaconda3\lib\site-packages

The best recall score is 0.9597515527950311
... with parameters: {'n_estimators': 50, 'max_features': 'sqrt', 'max_depth': 10, 'criterion': 'entropy'}


In [22]:
#Grid for XGBoost Classifier

param_grid_xg = {  
    'max_depth': range (2, 10, 1),
    'n_estimators': [10,50,250,1000,2000],
    'learning_rate': [1.0,0.2,0.1, 0.01, 0.05],}   

rand_search_xg = RandomizedSearchCV(estimator = xgboost, param_distributions=param_grid_xg, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# XGBoost Classifier model fit for grid search
_ = rand_search_xg.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search_xg.best_score_}")
print(f"... with parameters: {rand_search_xg.best_params_}")

bestRecallXg = rand_search_xg.best_estimator_

C:\Users\scbis\anaconda3\lib\site-packages\sklearn\model_selection\_search.py:292: UserWarning: The total space of parameters 200 is smaller than n_iter=500. Running 200 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


Fitting 5 folds for each of 200 candidates, totalling 1000 fits
The best recall score is 0.9655072463768116
... with parameters: {'n_estimators': 250, 'max_depth': 4, 'learning_rate': 0.2}


In [52]:
#Grid for Gradient Boost Classifier

param_grid_gb = {  
    'min_samples_split': np.arange(1,20),  
    'min_samples_leaf': np.arange(1,12),
    'min_impurity_decrease': np.arange(0.0001, 0.01, 0.0005), 
    'loss': ['log_loss', 'deviance', 'exponential'],
    'criterion': ['friedman_mse', 'squared_error'],
    'max_depth': range (2, 10, 1),
    'n_estimators': [10,50,250,1000,2000],
    'learning_rate': [1.0,0.2,0.1, 0.01, 0.05],}   

rand_search_gb = RandomizedSearchCV(estimator = gboost, param_distributions=param_grid_gb, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# Gradient Boost Classifier model fit for grid search
_ = rand_search_gb.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search_gb.best_score_}")
print(f"... with parameters: {rand_search_gb.best_params_}")

bestRecallgb = rand_search_gb.best_estimator_

Fitting 5 folds for each of 500 candidates, totalling 2500 fits


C:\Users\scbis\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
960 fits failed out of a total of 2500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
860 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\scbis\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\scbis\anaconda3\lib\site-packages\sklearn\ensemble\_gb.py", line 525, in fit
    self._check_params()
  File "C:\Users\scbis\anaconda3\lib\site-packages\sklearn\ensemble\_gb.py", line 282, in _check_params
    raise ValueError("Loss '{0:s}' not supported. ".format(self.loss))
ValueError: 

The best recall score is 0.9655072463768116
... with parameters: {'n_estimators': 250, 'min_samples_split': 15, 'min_samples_leaf': 7, 'min_impurity_decrease': 0.0041, 'max_depth': 5, 'loss': 'deviance', 'learning_rate': 0.2, 'criterion': 'friedman_mse'}


In [53]:
print("\n=========================LOGISTIC REGRESSION====================\n")
print(f"The best {score_measure} score is {rand_search_logr.best_score_}")
print(f"... with parameters: {rand_search_logr.best_params_}")
print("\n=========================DECISION TREE==========================\n")
print(f"The best {score_measure} score is {rand_search_tree.best_score_}")
print(f"... with parameters: {rand_search_tree.best_params_}")
print("\n==============================SVM===============================\n")
print(f"The best {score_measure} score is {rand_search_svm.best_score_}")
print(f"... with parameters: {rand_search_svm.best_params_}")
print("\n=========================ADABOOST===============================\n")
print(f"The best {score_measure} score is {rand_search_ada.best_score_}")
print(f"... with parameters: {rand_search_ada.best_params_}")
print("\n=========================RANDOMFOREST===========================\n")
print(f"The best {score_measure} score is {rand_search_rf.best_score_}")
print(f"... with parameters: {rand_search_rf.best_params_}")
print("\n=========================XGBOOST================================\n")
print(f"The best {score_measure} score is {rand_search_xg.best_score_}")
print(f"... with parameters: {rand_search_xg.best_params_}")
print("\n=========================GRADIENT BOOST================================\n")
print(f"The best {score_measure} score is {rand_search_gb.best_score_}")
print(f"... with parameters: {rand_search_gb.best_params_}")


=========================LOGISTIC REGRESSION====================

The best recall score is 0.9539958592132505
... with parameters: {'solver': 'liblinear', 'penalty': 'l1', 'max_iter': 200}

=========================DECISION TREE==========================

The best recall score is 0.9453416149068324
... with parameters: {'min_samples_split': 31, 'min_samples_leaf': 4, 'min_impurity_decrease': 0.0021, 'max_leaf_nodes': 43, 'max_depth': 15, 'criterion': 'entropy'}

==============================SVM===============================

The best recall score is 0.9712629399585921
... with parameters: {'kernel': 'linear', 'C': 1}

=========================ADABOOST===============================

The best recall score is 0.959792960662526
... with parameters: {'n_estimators': 1000, 'learning_rate': 0.1}

=========================RANDOMFOREST===========================

The best recall score is 0.9597515527950311
... with parameters: {'n_estimators': 50, 'max_features': 'sqrt', 'max_depth': 10, 'c

### Final Grid Search ###

In [29]:
# ADABoosting classifier grid
param_grid_ada = {  
     'n_estimators': [800,1000,1200],
     'learning_rate': [0.07,0.1,0.13],}  


grid_search_ada = GridSearchCV(estimator = adatree, param_grid=param_grid_ada, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# ADABoosting classifier fit
_ = grid_search_ada.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search_ada.best_score_}")
print(f"... with parameters: {grid_search_ada.best_params_}")

bestRecallAda = grid_search_ada.best_estimator_

Fitting 5 folds for each of 9 candidates, totalling 45 fits
The best recall score is 0.959792960662526
... with parameters: {'learning_rate': 0.07, 'n_estimators': 800}


In [31]:
#Grid for XGBoost Classifier

param_grid_xg = {  
    'max_depth': [2,4,6],
    'n_estimators': [200,250,300],
    'learning_rate': [0.17,0.2,0.23],}   

grid_search_xg = GridSearchCV(estimator = xgboost, param_grid=param_grid_xg, cv=kfolds,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# XGBoost Classifier model fit for grid search
_ = grid_search_xg.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search_xg.best_score_}")
print(f"... with parameters: {grid_search_xg.best_params_}")

bestRecallXg = grid_search_xg.best_estimator_

Fitting 5 folds for each of 27 candidates, totalling 135 fits
The best recall score is 0.9655072463768116
... with parameters: {'learning_rate': 0.2, 'max_depth': 4, 'n_estimators': 250}


In [32]:
#Grid for Randomforest Classifier
param_grid_rf = {  
     'n_estimators': [40,50,60],
     'max_features': ['sqrt'],
     'max_depth' : [8,10,12],
     'criterion' :['entropy'],}   

grid_search_rf = GridSearchCV(estimator = rforest, param_grid=param_grid_rf, cv=kfolds,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# ADABoost Classifier model fit for grid search
_ = grid_search_rf.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search_rf.best_score_}")
print(f"... with parameters: {grid_search_rf.best_params_}")

bestRecallRf = grid_search_rf.best_estimator_

Fitting 5 folds for each of 9 candidates, totalling 45 fits
The best recall score is 0.9568944099378882
... with parameters: {'criterion': 'entropy', 'max_depth': 10, 'max_features': 'sqrt', 'n_estimators': 50}


In [33]:
# Grid for SVM
param_grid_svm = {
    'C': [1,2,3],
    'kernel': ['linear'],  
}

grid_search_svm = GridSearchCV(estimator = svmm, param_grid=param_grid_svm, cv=kfolds,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# SVM model fit for grid search
_ = grid_search_svm.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search_svm.best_score_}")
print(f"... with parameters: {grid_search_svm.best_params_}")

bestRecallSvm = grid_search_svm.best_estimator_

Fitting 5 folds for each of 3 candidates, totalling 15 fits
The best recall score is 0.9712629399585921
... with parameters: {'C': 1, 'kernel': 'linear'}


In [35]:
# Grid for decision tree
 
param_grid_tree = {
    'min_samples_split': [29,31,33],  
    'min_samples_leaf': [2,4,6],
    'min_impurity_decrease': [0.0018,0.0021,0.0024],
    'max_leaf_nodes': [41,43,45], 
    'max_depth': [13,15,17], 
    'criterion': ['entropy'],
}

grid_search_tree = GridSearchCV(estimator = dtree, param_grid=param_grid_tree, cv=kfolds,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# Decision tree model fit for grid search
_ = grid_search_tree.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search_tree.best_score_}")
print(f"... with parameters: {grid_search_tree.best_params_}")

bestRecallTree = grid_search_tree.best_estimator_

Fitting 5 folds for each of 243 candidates, totalling 1215 fits
The best recall score is 0.94824016563147
... with parameters: {'criterion': 'entropy', 'max_depth': 13, 'max_leaf_nodes': 41, 'min_impurity_decrease': 0.0018, 'min_samples_leaf': 2, 'min_samples_split': 29}


In [36]:
#Grid for Logistic Regression

param_grid_logr = {
     'penalty': ['l1'],
     'solver': ['liblinear'],
     'max_iter': [180,200,220],
}

grid_search_logr = GridSearchCV(estimator = logreg, param_grid=param_grid_logr, cv=kfolds,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# Logistic Regression model fit for grid search
_ = grid_search_logr.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search_logr.best_score_}")
print(f"... with parameters: {grid_search_logr.best_params_}")

bestRecallLogr = grid_search_logr.best_estimator_

Fitting 5 folds for each of 3 candidates, totalling 15 fits
The best recall score is 0.9539958592132505
... with parameters: {'max_iter': 180, 'penalty': 'l1', 'solver': 'liblinear'}


In [55]:
#Grid for Gradient Boost Classifier

param_grid_gb = {  
    'min_samples_split': [13,15,17],  
    'min_samples_leaf': [5,7,9],
    'min_impurity_decrease': [0.0038,0.0041,0.0044], 
    'loss': ['deviance'],
    'criterion': ['friedman_mse'],
    'max_depth': [3,5,7],
    'n_estimators': [200,250,300],
    'learning_rate': [0.17,0.2,0.23],}   

grid_search_gb = GridSearchCV(estimator = gboost, param_grid=param_grid_gb, cv=kfolds,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

# Gradient Boost Classifier model fit for grid search
_ = grid_search_gb.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search_gb.best_score_}")
print(f"... with parameters: {grid_search_gb.best_params_}")

bestRecallgb = rand_search_gb.best_estimator_

Fitting 5 folds for each of 729 candidates, totalling 3645 fits
The best recall score is 0.9655072463768116
... with parameters: {'criterion': 'friedman_mse', 'learning_rate': 0.17, 'loss': 'deviance', 'max_depth': 7, 'min_impurity_decrease': 0.0038, 'min_samples_leaf': 5, 'min_samples_split': 17, 'n_estimators': 250}


## Final models with best parameters found ##

In [56]:
dtree = DecisionTreeClassifier(criterion='entropy', max_depth=13, max_leaf_nodes=41, min_impurity_decrease=0.0018, min_samples_leaf=2, min_samples_split=29)
svmm = SVC(C=1, kernel='linear')
logreg = LogisticRegression(max_iter=180, penalty='l1', solver='liblinear')
adatree = AdaBoostClassifier(learning_rate=0.07, n_estimators=800)
rforest = RandomForestClassifier(criterion='entropy', max_depth=10, max_features='sqrt', n_estimators=50)
xgboost = XGBClassifier(learning_rate=0.2, max_depth=4, n_estimators=250)
gboost = GradientBoostingClassifier(criterion='friedman_mse', learning_rate=0.17, loss='deviance', max_depth=7, min_impurity_decrease=0.0038, min_samples_leaf=5, min_samples_split=17, n_estimators=250)

## Model fit for train dataset and prediction with test dataset ##

In [40]:
_ = xgboost.fit(X_train, y_train)
y_pred = xgboost.predict(X_test)

print(f"{'Model':^18}{'Score':^18}")
print("************************************")
print(f"{'>> Recall Score:':18}{recall_score(y_test, y_pred)}")
print(f"{'Accuracy Score: ':18}{accuracy_score(y_test, y_pred)}")
print(f"{'Precision Score: ':18}{precision_score(y_test, y_pred)}")
print(f"{'F1 Score: ':18}{f1_score(y_test, y_pred)}")

xgboost_recall = recall_score(y_test, y_pred)

      Model             Score       
************************************
>> Recall Score:  0.959349593495935
Accuracy Score:   0.9887218045112782
Precision Score:  0.9915966386554622
F1 Score:         0.9752066115702479


In [41]:
_ = rforest.fit(X_train, y_train)
y_pred = rforest.predict(X_test)

print(f"{'Model':^18}{'Score':^18}")
print("************************************")
print(f"{'>> Recall Score:':18}{recall_score(y_test, y_pred)}")
print(f"{'Accuracy Score: ':18}{accuracy_score(y_test, y_pred)}")
print(f"{'Precision Score: ':18}{precision_score(y_test, y_pred)}")
print(f"{'F1 Score: ':18}{f1_score(y_test, y_pred)}")

rforest_recall = recall_score(y_test, y_pred)

      Model             Score       
************************************
>> Recall Score:  0.967479674796748
Accuracy Score:   0.9906015037593985
Precision Score:  0.9916666666666667
F1 Score:         0.9794238683127573


In [42]:
_ = adatree.fit(X_train, y_train)
y_pred = adatree.predict(X_test)

print(f"{'Model':^18}{'Score':^18}")
print("************************************")
print(f"{'>> Recall Score:':18}{recall_score(y_test, y_pred)}")
print(f"{'Accuracy Score: ':18}{accuracy_score(y_test, y_pred)}")
print(f"{'Precision Score: ':18}{precision_score(y_test, y_pred)}")
print(f"{'F1 Score: ':18}{f1_score(y_test, y_pred)}")

adatree_recall = recall_score(y_test, y_pred)

      Model             Score       
************************************
>> Recall Score:  0.983739837398374
Accuracy Score:   0.9962406015037594
Precision Score:  1.0
F1 Score:         0.9918032786885246


In [43]:
_ = logreg.fit(X_train, y_train)
y_pred = logreg.predict(X_test)

print(f"{'Model':^18}{'Score':^18}")
print("************************************")
print(f"{'>> Recall Score:':18}{recall_score(y_test, y_pred)}")
print(f"{'Accuracy Score: ':18}{accuracy_score(y_test, y_pred)}")
print(f"{'Precision Score: ':18}{precision_score(y_test, y_pred)}")
print(f"{'F1 Score: ':18}{f1_score(y_test, y_pred)}")

logreg_recall = recall_score(y_test, y_pred)

      Model             Score       
************************************
>> Recall Score:  0.975609756097561
Accuracy Score:   0.9906015037593985
Precision Score:  0.9836065573770492
F1 Score:         0.9795918367346939


In [44]:
_ = svmm.fit(X_train, y_train)
y_pred = svmm.predict(X_test)

print(f"{'Model':^18}{'Score':^18}")
print("************************************")
print(f"{'>> Recall Score:':18}{recall_score(y_test, y_pred)}")
print(f"{'Accuracy Score: ':18}{accuracy_score(y_test, y_pred)}")
print(f"{'Precision Score: ':18}{precision_score(y_test, y_pred)}")
print(f"{'F1 Score: ':18}{f1_score(y_test, y_pred)}")

svmm_recall = recall_score(y_test, y_pred)

      Model             Score       
************************************
>> Recall Score:  0.975609756097561
Accuracy Score:   0.9943609022556391
Precision Score:  1.0
F1 Score:         0.9876543209876543


In [45]:
_ = dtree.fit(X_train, y_train)
y_pred = dtree.predict(X_test)

print(f"{'Model':^18}{'Score':^18}")
print("************************************")
print(f"{'>> Recall Score:':18}{recall_score(y_test, y_pred)}")
print(f"{'Accuracy Score: ':18}{accuracy_score(y_test, y_pred)}")
print(f"{'Precision Score: ':18}{precision_score(y_test, y_pred)}")
print(f"{'F1 Score: ':18}{f1_score(y_test, y_pred)}")

dtree_recall = recall_score(y_test, y_pred)

      Model             Score       
************************************
>> Recall Score:  0.975609756097561
Accuracy Score:   0.9868421052631579
Precision Score:  0.967741935483871
F1 Score:         0.97165991902834


In [57]:
_ = gboost.fit(X_train, y_train)
y_pred = gboost.predict(X_test)

print(f"{'Model':^18}{'Score':^18}")
print("************************************")
print(f"{'>> Recall Score:':18}{recall_score(y_test, y_pred)}")
print(f"{'Accuracy Score: ':18}{accuracy_score(y_test, y_pred)}")
print(f"{'Precision Score: ':18}{precision_score(y_test, y_pred)}")
print(f"{'F1 Score: ':18}{f1_score(y_test, y_pred)}")

gboost_recall = recall_score(y_test, y_pred)

      Model             Score       
************************************
>> Recall Score:  0.975609756097561
Accuracy Score:   0.9943609022556391
Precision Score:  1.0
F1 Score:         0.9876543209876543


In [58]:
print("Recall scores...")
print(f"{'Decision Tree:':25}{dtree_recall}")
print(f"{'SVM:':25}{svmm_recall}")
print(f"{'Logistic Regression':25}{logreg_recall}")

print(f"{'Random Forest:':25}{rforest_recall}")
print(f"{'Ada Boosted Tree:':25}{adatree_recall}")
print(f"{'XGBoost Tree:':25}{xgboost_recall}")
print(f"{'Gradient Boosting':25}{gboost_recall}")

Recall scores...
Decision Tree:           0.975609756097561
SVM:                     0.975609756097561
Logistic Regression      0.975609756097561
Random Forest:           0.967479674796748
Ada Boosted Tree:        0.983739837398374
XGBoost Tree:            0.959349593495935
Gradient Boosting        0.975609756097561


## Analysis of Models ##

The analysis of recall score of models show that ADABoost Tree is the best model (0.984). 
While XGBoost performs worst (0.96), the Decision Tree, SVM, Logistic Regression, and Gradient Boosting perform same (0.976). The random forest performs better than XGBoost (.967), but not better than most of the models.

This shows that the esemble models perform better than the pruning models.